server/plugin.go

```diff
@@ -83,43 +74,77 @@ func (p *Plugin) OnActivate() error {
-	if err != nil {
-		return fmt.Errorf("could not parse the listen address %q", *appCfg.MetricsSettings.ListenAddress)
-	}
-	if host == "" {
-		host = "localhost"
+	manager := scrape.NewManager(nil, p.logger, p.db)
+	syncCh := make(chan map[string][]*targetgroup.Group)
+	p.closeChan = make(chan bool)
+	p.waitGroup = sync.WaitGroup{}
+
+	// we start the manager first, then apply the scrape config
+	p.waitGroup.Add(1)
+	go func() {
+		defer p.waitGroup.Done()
+
+		p.API.LogInfo("Running scrape manager...")
+		err2 := manager.Run(syncCh)
+		if err2 != nil {
+			p.API.LogError("scrape manager exited unexpectedly", "err", err2)
+		}
+	}()
+
+	scpCfg := &config.Config{
+		ScrapeConfigs: []*config.ScrapeConfig{
+			{
+				JobName:                    "prometheus",
+				Scheme:                     "http",
+				MetricsPath:                "metrics",
+				ScrapeInterval:             model.Duration(time.Duration(*p.configuration.ScrapeIntervalSeconds) * time.Second),
+				ScrapeTimeout:              model.Duration(time.Duration(*p.configuration.ScrapeTimeoutSeconds) * time.Second),
+				BodySizeLimit:              units.Base2Bytes(*p.configuration.BodySizeLimitBytes),
+				HonorLabels:                *p.configuration.HonorTimestamps,
+				SampleLimit:                uint(*p.configuration.SampleLimit),
+				NativeHistogramBucketLimit: uint(*p.configuration.BucketLimit),
+			},
+		},
 	}
+	manager.ApplyConfig(scpCfg)
 
-	// TODO(isacikgoz): Use multiple targets for HA env
-	lb := labels.NewBuilder(labels.FromMap(map[string]string{
-		model.AddressLabel:        host + ":" + port, // this is used for labeling the source
-		model.ScrapeIntervalLabel: fmt.Sprintf("%ds", scrapeInterval),
-		model.ScrapeTimeoutLabel:  fmt.Sprintf("%ds", *p.configuration.ScrapeTimeoutSeconds),
-	}))
-
-	lset, origLabels, err := scrape.PopulateLabels(lb, &config.ScrapeConfig{
-		JobName: "prometheus",
-	}, true)
+	sync, err := generateTargetGroup(p.API.GetConfig(), nil)
 	if err != nil {
-		return err
+		return fmt.Errorf("could not set scrape target :%w", err)
 	}
+	syncCh <- sync
 
-	target := scrape.NewTarget(lset, origLabels, url.Values{})
-
-	// Mutator is being used to apply labels to the metric samples
-	sampleMutator = func(l labels.Labels) labels.Labels {
-		return mutateSampleLabels(l, target, *p.configuration.HonorTimestamps, []*relabel.Config{})
+	// check if cluster is enabled
+	if p.isHA() {
+		// TODO(isacikgoz): get cluster info
+		// we will need to push new cluster layout to p.clusterCh by either polling the cluster table
+		// or listening the cluster event messages
+		p.API.LogWarn("cluster meterics is not enabled")
 	}
 
+	// this goroutine will need to be re-structurd to listen a more channels
+	// once we start supporting HA, we will need to listen the cluster change channel and
+	// convert the []mmodel.ClusterDiscovery entries into map[string][]*targetgroup.Group
+	p.waitGroup.Add(1)
+	go func() {
```

Comment:
Just for clarity, you don't really need a separate goroutine at this point, but we are creating it for when we will need HA?

Reply:
Exactly, there will be `select` to listen different channels for HA. Therefore I added the comment above.